packages/remix-react/scroll-restoration.tsx

```diff
@@ -1,49 +1,62 @@
 import * as React from "react";
-import { useLocation } from "react-router-dom";
+import type { ScrollRestorationProps as ScrollRestorationPropsRR } from "react-router-dom";
+import {
+  useLocation,
+  UNSAFE_useScrollRestoration as useScrollRestoration,
+} from "react-router-dom";
 
-import { useBeforeUnload, useTransition } from "./components";
 import type { ScriptProps } from "./components";
+import { useMatches } from "./components";
 
 let STORAGE_KEY = "positions";
-
-let positions: { [key: string]: number } = {};
-
-if (typeof document !== "undefined") {
-  let sessionPositions = sessionStorage.getItem(STORAGE_KEY);
-  if (sessionPositions) {
-    positions = JSON.parse(sessionPositions);
-  }
-}
+let hydrated = false;
 
 /**
  * This component will emulate the browser's scroll restoration on location
  * changes.
  *
  * @see https://remix.run/api/remix#scrollrestoration
  */
-export function ScrollRestoration(props: ScriptProps) {
-  useScrollRestoration();
-
-  // wait for the browser to restore it on its own
-  React.useEffect(() => {
-    window.history.scrollRestoration = "manual";
-  }, []);
-
-  // let the browser restore on it's own for refresh
-  useBeforeUnload(
-    React.useCallback(() => {
-      window.history.scrollRestoration = "auto";
-    }, [])
+export function ScrollRestoration({
+  getKey,
+  ...props
+}: ScriptProps & {
+  getKey: ScrollRestorationPropsRR["getKey"];
+}) {
+  let location = useLocation();
+  let matches = useMatches();
+
+  useScrollRestoration({
+    getKey,
+    storageKey: STORAGE_KEY,
+    skip: !hydrated,
```

Comment:
Matt to so some double checking here and see if there's any way `skip` could go away...

Reply:
If we assume `hydrationData` means "skip the first restoration" then we could get rid of this